File: lintcode/construct_binary_tree_from_preorder_and_inorder_traversal.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""


class Solution:
    """
    @param preorder : A list of integers that preorder traversal of a tree
    @param inorder : A list of integers that inorder traversal of a tree
    @return : Root of a tree
    """
    def buildTree(self, preorder, inorder):
        if inorder:
            root = TreeNode(preorder.pop(0))
            index = inorder.index(root.val)
            right, inorder[index:] = inorder[index + 1:], []
            root.left, root.right = self.buildTree(preorder, inorder), \
                                    self.buildTree(preorder, right)
            return root

File: leetcode/construct_binary_tree_from_preorder_and_inorder_traversal.js
/* Given preorder and inorder traversal of a tree, construct the binary tree.
 * Time Complexity: O(n)
 * Space Complexity: O(n), O(h) is achievable assuming partition in-place
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function(preorder, inorder) {
    if (inorder.length > 0) {
        var root = new TreeNode(preorder.shift());
        var index = inorder.indexOf(root.val);
        var right = inorder.splice(index + 1, inorder.length - index - 1);
        inorder.pop();
        root.left = buildTree(preorder, inorder);
        root.right = buildTree(preorder, right);
        return root;
    }
    return null;
};
